File: probe/_replay_worker.py
"""Replay worker: load ONE version of a module (package-aware, from a git
worktree) and run a function/method over captured arguments, emitting canonical
observations. Runs as a subprocess so two versions of the same package never
share an interpreter.

stdin  JSON: {worktree, module_name, qualname, args_b64: [...]}
stdout JSON: {"loaded": bool, "error": str|null, "obs": [ {val|exc|nondet, io, threads}, ... ]}

qualname may be dotted ("ClassName.method"). For an instance method the captured
args[0] is `self`, which pickle rebuilds as THIS version's class (the worktree is
on sys.path), so the method runs against the right version. Classmethods (bound)
get their captured `cls` dropped.
"""

import base64
import inspect
import json
import os
import pickle
import sys


def _resolve(module, qualname):
    obj = module
    for part in qualname.split("."):
        obj = getattr(obj, part)
    return obj


def main() -> int:
    job = json.load(sys.stdin)
    out = {"loaded": False, "error": None, "obs": []}
    try:
        # Support both flat (pkg/) and src (src/pkg/) layouts; insert at front so
        # the worktree version shadows any installed copy of the package.
        wt = job["worktree"]
        for p in (os.path.join(wt, "src"), wt):
            if os.path.isdir(p):
                sys.path.insert(0, p)
        import importlib

        from probe import harness
        from probe.canonical import canonical

        module = importlib.import_module(job["module_name"])
        try:
            fn = _resolve(module, job["qualname"])
        except AttributeError:
            # module imported fine but the function isn't here: it was added or
            # removed across versions. Report 'absent' (not an error) so the
            # verdict can be 'skipped', not a spurious failure.
            out["absent"] = True
            print(json.dumps(out))
            return 0
        if not callable(fn):
            out["error"] = "not callable in this version"
            print(json.dumps(out))
            return 0
        out["loaded"] = True
        try:
            out["params"] = list(inspect.signature(fn).parameters)
        except (TypeError, ValueError):
            out["params"] = None
        import copy
        bound_classmethod = inspect.ismethod(fn)  # cls already bound
        qualname = job["qualname"]
        is_method = ("." in qualname) and not bound_classmethod

        def run_once(values):
            # Deep-copy so a mutating call doesn't taint the next run, and so we
            # can read the receiver's post-call state. Behavior of a method =
            # return value PLUS how it mutated self.
            try:
                args = copy.deepcopy(values)
            except Exception:
                args = values
            call = tuple(args[1:]) if bound_classmethod else tuple(args)
            o = harness.observe(fn, call)
            ret = ("exc", o.exception) if o.exception is not None \
                else ("val", canonical(o.value))
            state = None
            if is_method and args:
                try:
                    state = canonical(args[0])
                except Exception:
                    state = ["opaque", "self"]
            return ret, state, o.counts.get("io", 0), o.counts.get("threads", 0)

        for blob in job["args_b64"]:
            values = pickle.loads(base64.b64decode(blob))
            r1, s1, io, th = run_once(values)
            r2, s2, _, _ = run_once(values)  # determinism guard (fresh copy)
            rec = {"io": io, "threads": th}
            if r1 != r2 or s1 != s2:
                rec["nondet"] = True
            else:
                if r1[0] == "exc":
                    rec["exc"] = r1[1]
                else:
                    rec["val"] = r1[1]
                if is_method:
                    rec["self_after"] = s1
            out["obs"].append(rec)
    except Exception as e:  # import / relative-import / unpickle error
        out["error"] = "%s: %s" % (type(e).__name__, e)
    print(json.dumps(out))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: packages/node/src/replayWorker.js
'use strict';
/**
 * Replay worker: load ONE version of a module (from a given version root) and
 * run a function/method over captured arguments, emitting canonical
 * observations. Runs as a subprocess so two versions never share a runtime.
 * Mirrors packages/python/probe/_replay_worker.py.
 *
 * stdin  JSON: {versionRoot, moduleRel, qualname, is_method, args_b64:[...]}
 * stdout JSON: {loaded, error, absent, params, obs:[{val|exc, self_after?, io, threads, nondet?}]}
 */

const path = require('node:path');
const v8 = require('node:v8');
const { observe, canonical } = require('./harness');
const { deepEqual } = require('./soundness');

function readStdin() {
  return new Promise((resolve) => {
    let data = '';
    process.stdin.setEncoding('utf8');
    process.stdin.on('data', (c) => { data += c; });
    process.stdin.on('end', () => resolve(data));
  });
}

function resolveTarget(mod, qualname) {
  // "foo" -> {fn: mod.foo}; "Class.method" -> {fn: proto.method, isMethod}
  if (qualname.includes('.')) {
    const [cls, method] = qualname.split('.');
    const klass = mod[cls];
    if (!klass || !klass.prototype) return null;
    const fn = klass.prototype[method];
    return typeof fn === 'function' ? { fn, method: true } : null;
  }
  const fn = mod[qualname];
  return typeof fn === 'function' ? { fn, method: false } : null;
}

async function runOnce(target, isMethod, values) {
  let callThis = null;
  let callArgs = values;
  if (isMethod) {
    callThis = values[0];
    callArgs = values.slice(1);
  }
  const bound = isMethod ? (...a) => target.fn.apply(callThis, a) : (...a) => target.fn(...a);
  const o = await observe(bound, callArgs);
  const ret = o.exception !== null ? ['exc', o.exception] : ['val', canonical(o.value)];
  let state = null;
  if (isMethod) {
    try { state = canonical(callThis); } catch (e) { state = ['opaque', 'self', '<unrepresentable>']; }
  }
  return { ret, state, io: o.counts.io, threads: o.counts.threads };
}

async function main() {
  const out = { loaded: false, error: null, obs: [] };
  try {
    const job = JSON.parse(await readStdin());
    const modPath = path.join(job.versionRoot, job.moduleRel);
    delete require.cache[require.resolve(modPath)];
    const mod = require(modPath);

    const target = resolveTarget(mod, job.qualname);
    if (!target) { out.absent = true; process.stdout.write(JSON.stringify(out)); return; }
    out.loaded = true;
    const isMethod = !!job.is_method && target.method;
    out.params = (target.fn.length != null) ? target.fn.length : null;

    for (const b64 of job.args_b64) {
      let values;
      try {
        values = v8.deserialize(Buffer.from(b64, 'base64'));
      } catch (e) { out.obs.push({ val: ['opaque', 'unpicklable', '<unrepresentable>'], io: 0, threads: 0 }); continue; }
      const r1 = await runOnce(target, isMethod, values);
      const r2 = await runOnce(target, isMethod, values); // determinism guard
      const rec = { io: r1.io, threads: r1.threads };
      if (!deepEqual(r1.ret, r2.ret) || !deepEqual(r1.state, r2.state)) {
        rec.nondet = true;
      } else {
        if (r1.ret[0] === 'exc') rec.exc = r1.ret[1];
        else rec.val = r1.ret[1];
        if (isMethod) rec.self_after = r1.state;
      }
      out.obs.push(rec);
    }
  } catch (e) {
    out.error = `${e && e.constructor ? e.constructor.name : 'Error'}: ${e && e.message}`;
  }
  process.stdout.write(JSON.stringify(out));
}

main();
